internal/pkg/consensus/expected.go

```diff
@@ -289,7 +287,7 @@ func (c *Expected) runMessages(ctx context.Context, st state.Tree, vms vm.Storag
 	// NOTE: It is possible to optimize further by applying block validation
 	// in sorted order to reuse first block transitions as the starting state
 	// for the tipSetProcessor.
-	allMessages := combineMessages(blsMessages, secpMessages)
+	allMessages := append(blsMessages, secpMessages...)
```

Comment:
This is wrong. I realized its wrong after I approved. It's wrong because it replicates incorrect behavior I introduced with BLS messages. ProcessTipSet expects a slice of slice of messages where each inner slice corresponds to a block. This is creating a slice of slices that is twice as long as the number of blocks. It's troubling that this is not being causing tests to fail.

I've fixed this in my pending PR, so it's ok to leave this broken here.

I need to examine to spec to determine whether tipset message processing should be ordered by block, type, gas price, or type, block, gas price. If the latter, we're going to need to pass in a different data structure that lets us tie a block miner to an arbitrary message.

Reply:
Thanks, I will leave this broken here.

It would seem worthwhile to be for us to create an explicit type representing the messages in a tipset, encapsulating the ordering and de-duping logic. As a bonus, consider arranging it so that it unavoidably checks all the message signature while extracting them from the message, thereby giving some type-safety to signature validation.